Migrate to System.Text.Encoding.Web

Tratcher opened the pull request:
#391 @davidfowl @muratg @Eilon 
David asked me to keep the interfaces for DI, so I replaced the encoders with wrappers. Components that were not using the interfaces I change to directly use the BCL types.

Tratcher commented on src/Microsoft.Extensions.WebEncoders.Core/HtmlEncoderWrapper.cs:
```diff
@@ -0,0 +1,45 @@
+            return _encoder.Encode(value);
+        }
+
+        public void HtmlEncode(string value, int startIndex, int charCount, TextWriter output)
```
Should we change these method names and parameter ordering to match the new BCL APIs? It will make it easier for people to switch between our DI wrapper and the underlying BCL types, but it's more of a break for us.

muratg replied:
Having parity between the wrapper and BCL types makes sense to me. How much work would it be to do it?

Tratcher commented:
@muratg Updated with the new names and signatures. It will make the down stream breaks much worse, but is probably better in the long run.

Tratcher commented:
+@rynowak who does not like wrappers.

muratg commented:
Looks good to me.

Tratcher commented:
@davidfowl Removing the wrappers mostly works, except that it's much harder to mock and test because the needed methods aren't virtual.
https://github.com/aspnet/HttpAbstractions/commit/90c13adc5d90108a4826af3b5149a3e923b79436
Compare the old CommonTestEncoder (https://github.com/aspnet/HttpAbstractions/commit/90c13adc5d90108a4826af3b5149a3e923b79436#diff-5093aee64e24c262ac1165902a6ceef9) with the new HtmlTestEncoder (https://github.com/aspnet/HttpAbstractions/commit/90c13adc5d90108a4826af3b5149a3e923b79436#diff-387f3dc26a94ebcb7c37dd754bab84c5)

Tratcher commented:
https://github.com/dotnet/corefx/pull/4038

Eilon commented on src/Microsoft.AspNet.Http.Abstractions/project.json:
```diff
@@ -11,13 +11,19 @@
     "dnxcore50": {
       "dependencies": {
         "System.Collections": "4.0.11-beta-*",
+        "System.ComponentModel": "4.0.1-beta-*",
```
Hmm, is this actually used? I'm curious for what...

Tratcher replied:
System.Text.Encodings.Web seems to require it.

Eilon replied:
Won't the transitive dependency solve that?

davidfowl replied:
Should...

Eilon commented:
:shipit:

muratg commented on src/Microsoft.AspNet.Http.Abstractions/QueryString.cs:
```diff
@@ -154,13 +163,23 @@ public static QueryString Create(IEnumerable<KeyValuePair<string, StringValues>>
             {
+                if (pair.Key == null)
+                {
+                    throw new ArgumentNullException(nameof(pair.Key));
```
`InvalidOperationException` as this is not an arg.

muratg commented on src/Microsoft.AspNet.Http.Abstractions/QueryString.cs:
```diff
@@ -154,13 +163,23 @@ public static QueryString Create(IEnumerable<KeyValuePair<string, StringValues>>
                 {
+                    if (value == null)
+                    {
+                        throw new ArgumentNullException(nameof(value));
```
Ditto.

Eilon commented:
Clarification: My :shipit: was contingent on an offline discussion regarding keeping the interfaces. I don't think it's bad either way, but I prefer the nice interfaces.

davidfowl commented on src/Microsoft.Extensions.WebEncoders.Core/project.json:
```diff
@@ -1,27 +1,33 @@
+  "dependencies": {
+    "System.Text.Encodings.Web": "4.0.0-beta-*"
+  },
+  "frameworks": {
```
These need to be updated to use generations now

davidfowl commented on src/Microsoft.Extensions.WebEncoders.Core/UrlEncoderWrapper.cs:
```diff
@@ -0,0 +1,46 @@
+            _encoder = UrlEncoder.Create(settings);
+        }
+
+        public static IUrlEncoder Default { get; } = new UrlEncoderWrapper(UrlEncoder.Default);
```
Do we still need Default? People using this should just use the BCL type right?

Tratcher replied:
This default is for plugging into places that take the interface, see https://github.com/aspnet/HttpAbstractions/pull/449/files#diff-af79926afbc0631c7212bff6ed282bc0R208

Tratcher commented:
Rebased and updated. Now that we have the right virtuals we can remove the interfaces and the entire WebEncoders.Core package.

CommonTestEncoder is also replaced by HtmlTestEncoder, UrlTestEncoder, and JavaScriptTestEncoder, provided directly in the WebEncoders package.

Eilon commented:
I miss the interfaces, but :shipit:
